model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,657 @@
+
+        LOG.tracef("removeUserSession(%s, %s)%s", realm, session, getShortStackTrace());
+
+        userSessionTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void removeUserSessions(RealmModel realm, UserModel user) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = userSessionStore.createCriteriaBuilder()
+                .compare(UserSessionModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserSessionModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, user.getId());
+
+        LOG.tracef("removeUserSessions(%s, %s)%s", realm, user, getShortStackTrace());
+
+        userSessionTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void removeAllExpired() {
+        session.realms().getRealmsStream().forEach(this::removeExpired);
+    }
+
+    @Override
+    public void removeExpired(RealmModel realm) {
+        int currentTime = Time.currentTime();
+        int expired = currentTime - realm.getSsoSessionMaxLifespan();
+        int expiredRefresh = currentTime - realm.getSsoSessionIdleTimeout() - SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+        int expiredRememberMe = currentTime - (realm.getSsoSessionMaxLifespanRememberMe() > 0 ?
+                realm.getSsoSessionMaxLifespanRememberMe() : realm.getSsoSessionMaxLifespan());
+        int expiredRefreshRememberMe = currentTime - (realm.getSsoSessionIdleTimeoutRememberMe() > 0 ?
+                realm.getSsoSessionIdleTimeoutRememberMe() : realm.getSsoSessionIdleTimeout()) -
+                SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+        int expiredOffline = currentTime - realm.getOfflineSessionIdleTimeout() - SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+
+        // remove expired user sessions and its client sessions
+        ModelCriteriaBuilder<UserSessionModel> userSessionMcb = userSessionStore.createCriteriaBuilder()
+                .compare(UserSessionModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserSessionModel.SearchableFields.IS_EXPIRED, ModelCriteriaBuilder.Operator.EQ,
+                        expiredRememberMe, expiredRefreshRememberMe, expired, expiredRefresh, expiredOffline);
+
+        userSessionTx.delete(UUID.randomUUID(), userSessionMcb);
+    }
+
+    @Override
+    public void removeUserSessions(RealmModel realm) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false);
+
+        LOG.tracef("removeUserSessions(%s)%s", realm, getShortStackTrace());
+
+        userSessionTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void onRealmRemoved(RealmModel realm) {
+        removeUserSessions(realm);
+        session.loginFailures().removeAllUserLoginFailures(realm);
```

Comment:
Should this line remain here?

Reply:
Moved to the equivalent event listener in the `MapUserLoginFailureProviderFactory`.

Reply:
I don't think it is there